graph.go

```diff
@@ -4,193 +4,166 @@
-// this function will take precedence over the Uniform Cost function (all weights are 1) if "nil"
-// is passed in for the function argument.
-//
-// If the argument is nil, or the edge is invalid for some reason, this should return math.Inf(1)
-type Coster interface {
-	Cost(Edge) float64
+	// EdgeBetween returns the edge between nodes u and v.
+	EdgeBetween(u, v Node) Edge
 }
 
-type CostGraph interface {
-	Coster
+// Directed is a directed graph.
+type Directed interface {
 	Graph
+
+	// EdgeFromTo returns the edge leading from u to v.
+	EdgeFromTo(u, v Node) Edge
+
+	// To returns all nodes that can be lead to the
+	// given node.
+	To(Node) []Node
 }
 
-type CostDirectedGraph interface {
-	Coster
-	DirectedGraph
+// EdgeLister wraps the Edges method.
+type EdgeLister interface {
+	Edges() []Edge
 }
 
-// A graph that implements HeuristicCoster implements a heuristic between any two given nodes.
-// Like Coster, if a graph implements this and a function needs a heuristic cost (e.g. A*), this
-// function will take precedence over the Null Heuristic (always returns 0) if "nil" is passed in
-// for the function argument. If HeuristicCost is not intended to be used, it can be implemented as
-// the null heuristic (always returns 0).
-type HeuristicCoster interface {
-	// HeuristicCost returns a heuristic cost between any two nodes.
-	HeuristicCost(n1, n2 Node) float64
+// Weighter wraps the Weight method.
+type Weighter interface {
+	// Weight returns the edge weight for the parameter,
+	Weight(Edge) float64
 }
 
-// A Mutable is a graph that can have arbitrary nodes and edges added or removed.
-//
-// Anything implementing Mutable is required to store the actual argument. So if AddNode(myNode) is
-// called and later a user calls on the graph graph.NodeList(), the node added by AddNode must be
-// an the exact node, not a new node with the same ID.
-//
-// In any case where conflict is possible (e.g. adding two nodes with the same ID), the later
-// call always supercedes the earlier one.
-//
-// Functions will generally expect one of MutableGraph or MutableDirectedGraph and not Mutable
-// itself. That said, any function that takes Mutable[x], the destination mutable should
-// always be a different graph than the source.
+// Mutable wraps generalized graph alteration methods.
 type Mutable interface {
-	// NewNode returns a node with a unique arbitrary ID.
-	NewNode() Node
+	// NewNodeID returns a new unique arbitrary ID.
+	NewNodeID() int
 
 	// Adds a node to the graph. AddNode panics if
 	// the added node ID matches an existing node ID.
 	AddNode(Node)
 
-	// RemoveNode removes a node from the graph, as well as any edges
-	// attached to it. If no such node exists, this is a no-op, not an error.
+	// RemoveNode removes a node from the graph, as
+	// well as any edges attached to it. If the node
+	// is not in the graph it is a no-op.
 	RemoveNode(Node)
-}
```

Comment:
Do we need a "ReplaceNode"? This way you could swap a set of information without deleting all of the edges.